rplidar.go

```diff
@@ -97,64 +102,87 @@ func newRplidar(ctx context.Context, _ resource.Dependencies, c resource.Config,
 
 	logger.Info("found and connected to an " + rplidarModelByteMap[rplidarDevice.model] + " rplidar")
 
-	rp := &Rplidar{
-		Named:                   c.ResourceName().AsNamed(),
-		device:                  rplidarDevice,
-		nodeSize:                8192,
-		logger:                  logger,
-		defaultNumScans:         1,
-		warmupNumDiscardedScans: 5,
-		minRangeMM:              svcConf.MinRangeMM,
+	rp := &rplidar{
+		Named:      c.ResourceName().AsNamed(),
+		device:     rplidarDevice,
+		minRangeMM: svcConf.MinRangeMM,
+
+		cacheMutex:             sync.RWMutex{},
+		cacheBackgroundWorkers: sync.WaitGroup{},
+
+		logger: logger,
+	}
+
+	// Setup RPLiDAR
+	if err := rp.setupRPLidar(ctx); err != nil {
+		return nil, errors.Wrap(err, "there was a problem setting up the rplidar")
 	}
 
-	rp.mu.Lock()
-	defer rp.mu.Unlock()
-	rp.started = true
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	rp.cancelFunc = cancelFunc
 
-	// S1 rplidars do not require the motor to be started before scanning can begin
+	// Start background caching of pointcloud data
+	rp.cacheBackgroundWorkers.Add(1)
+	go func() {
+		defer rp.cacheBackgroundWorkers.Done()
+		rp.cachePointCloudLoop(cancelCtx)
+	}()
+
+	return rp, nil
+}
+
+// setupRPLiDAR starts the motor, if necessary, and warms up the device, discard several scans in the
+// process to ensure data returned to the user is valid.
```

Comment:
```suggestion
// setupRPLiDAR starts the motor, if necessary, warms up the device, and discards several scans in the
// process to ensure data returned to the user is valid.
```

Reply:
Updated comment, slightly different than suggested but let me know if that works for you